docs/nativeaot.md

```diff
@@ -0,0 +1,49 @@
+# NativeAOT
+
+We've added *experimental* support for using [NativeAOT][1] when publishing iOS,
+tvOS, Mac Catalyst and macOS apps in .NET 8.
+
+NativeAOT may produce smaller and/or faster apps - or it may not. It's very
+important test and profile to determine the results of enabling NativeAOT.
+
+NativeAOT is enabled by setting the `PublishAot` property to `true` in the project file:
```

Comment:
It's not necessary to set `$(PublishAotUsingRuntimepack)=true` anymore?

Reply:
Afaik it is still required if you do not want to change your project file but just force defaults through CLI, e.g.,:
```
dotnet publish -c Release -f net8.0-ios -r ios-arm64 -p:PublishAot=true -p:PublishAotUsingRuntimePack=true
```

However, passing `-p:PublishAot=true` on the command-line is discouraged in general, as it will not just enable NativeAOT-targets, but has some additional dependencies in the SDK to run analyzers, disable generating runtimeconfig.json,...  So I suppose it is OK that we do not mention it and avoid confusion.

Regarding the project-level setting, Xamarin SDK helps by setting the property for the user:
https://github.com/xamarin/xamarin-macios/blob/2534418caa473fc7e869095e220329230866ff41/dotnet/targets/Xamarin.Shared.Sdk.props#L179-L185

Reply:
I'm not sure I understand why it would be different when passing PublishAot from the command line vs in the project?

Reply:
> I'm not sure I understand why it would be different when passing PublishAot from the command line vs in the project?

There are issues with the framework resolution in the SDK, when we have multiple target frameworks specified in the project file (e.g., MAUI projects). 
Building a MAUI app without adding `PublishAot` to the project file, but passing it through command line:

```
dotnet publish -f net8.0-ios -c Release -r ios-arm64 -p:PublishAot=true
```

Gives the following error - coming from the SDK:
```
/usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1203: Ahead-of-time compilation is not supported for the target runtime identifier 'ios-arm64'. [/Users/ivan/tmp/AppTest/AppTest.csproj::TargetFramework=net8.0-android]
```

Including the `-p:PublishAotUsingRuntimePack=true` in the command line above, works around the issue.

--- 
Additionally, regarding what I said about passing `PublishAot` only through command line being "discouraged", is the fact that `dotnet build` and `dotnet publish` with and without `PublishAot=true` mean different things, which can be confusing to the end user. 
If we consider a desktop app doing the following:

```csharp
var asm = System.Reflection.Assembly.LoadFrom("somepath");
``` 
1. When built with `dotnet build` - just builds the app as a regular managed desktop app:
    - outputs:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 78 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Debug/net8.0/NaotApp.dll

    Build succeeded.
    0 Warning(s)
    0 Error(s)
    ```

2. When built with `dotnet build -p:PublishAot=true` - builds the app as a managed desktop app with trim/AOT analyzers enabled (indicating it might be published/deployed afterwards):
    - outputs:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 127 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    /Users/ivan/tmp/NaotApp/Program.cs(1,11): warning IL2026: Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Debug/net8.0/NaotApp.dll

    Build succeeded.

    /Users/ivan/tmp/NaotApp/Program.cs(1,11): warning IL2026: Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    1 Warning(s)
    0 Error(s)
    ```
3. When built with `dotnet publish` - deploys the app without enabling trim/AOT analyzers (although I am not 100% sure in which mode, I would suppose it is still a framework dependent one)
    - output:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 78 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy     [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/NaotApp.dll
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/publish/
    ```
4. When built with `dotnet publish -p:PublishAot=true` - deploys the app with enabled trim/AOT analyzers and with NativeAOT
    - output:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 149 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message     NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy     [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    /Users/ivan/tmp/NaotApp/Program.cs(1,11): warning IL2026: Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/NaotApp.dll
    Generating native code
    /Users/ivan/tmp/NaotApp/Program.cs(1): Trim analysis warning IL2026: Program.<Main>$(String[]): Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/publish/
    ```
5. Additionally, when built with `dotnet publish -p:PublishReadyToRun=true` - deploys the app without enabling trim/AOT analyzers and with ReadyToRunCrossgen2
    - output:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 9.55 sec).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message     NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy     [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/NaotApp.dll
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/publish/
    ```
    
Because of all these differences when using `dotnet build` or `dotnet publish` (and various deployment modes), I believe the principle is: `If you want to use NativeAOT - set the opt-in property in the project file`
This gives you: 
- necessary information (trim/AOT warnings) for the inner-dev loop experience - `dotnet build` to guide how you write your code so it is trim/NativeAOT compatible 
- on `dotnet publish` the deployment mode compatible with step 1. 
- finally, a simple way of using NativeAOT ie without needing to understand all other modes and concepts 

I know I gave an example for desktop platforms, but for mobile we also have similar confusion for Xamarin legacy / Xamarin .NET, MonoAOT / MonoAOT+interpreter, MonoAOT vs NativeAOT, so I suppose for simplicity we should try to follow the same principle, at least until we do not solve SDK issues like the need for `PublishAotUsingRuntimePack` and https://github.com/dotnet/sdk/issues/25392 

---

Hope this helps :)

Reply:
@ivanpovazan I think that we should focus just on the csproj settings here in this document. Does it mean that since we suggest adding `<PublishAot>true</PublishAot>` to the .csproj, we should also guide the devs to put `<PublishAotUsingRuntimepack>true</PublishAotUsingRuntimepack>` next to it?

Reply:
If `PublishAot` is specified in the project file, `PublishAotUsingRuntimepack` is not needed.